Checks the latest episodes on SCPR.org and compares them with audio files from Megaphone

arellaTV opened the pull request:
Main features:
- Can configure which programs to watch via environment variables
- Can configure what duration difference threshold via environment variables
- Gives a rundown of each episode's status at the end of each invocation, i.e. `airtalk's audio file is within an expected range`.
- Catches all errors and logs it at the end of each invocation

From my testing, it looks like the audio files will always have a different filesize and duration, even when manually uploading through Megaphone's CMS. because the biggest difference I saw was roughly 4.5 seconds, I defaulted the difference threshold to be 6 seconds (meaning, if the difference in duration between the two files is greater than 6, then something is wrong and a re-upload should be started).

I also added a `publish.sh` script for convenience in publishing the lambda.

arellaTV committed:
Reuploads if difference is too great

arellaTV committed:
Move directories around to match lambda format

arellaTV committed:
Abstract methods to their own files

arellaTV committed:
Promisify all library functions

arellaTV committed:
Requires dotenv for scripts that need env vars

arellaTV committed:
Adds more inline documentation for comparision script

arellaTV committed:
Adds publish script

nakedsushi commented on megaphone-validator/.env.example:
```diff
@@ -0,0 +1,4 @@
+DIFFERENCE_THRESHOLD=
+MEGAPHONE_AUTH_TOKEN=
+MEGAPHONE_NETWORK_ID=
+SCPR_PROGRAM_SLUGS=
```
nitpick: should be a newline at the end of this line.

nakedsushi commented on megaphone-validator/index.js:
```diff
@@ -0,0 +1,39 @@
+
+    // Exit if there are no podcasts indicated in the env
+    if (!programSlugString) {
+        return false;
```
If the point of entry is in this file, we should exit early if any of the required env variables like `MEGAPHONE_AUTH_TOKEN` or `MEGAPHONE_NETWORK_ID` are missing, since without those, this script wouldn't run to completion anyway, right?

arellaTV replied:
@nakedsushi Ah, yes you're right! I went ahead and added checks for those two as well so that the function can exit early if those aren't indicated

nakedsushi commented on megaphone-validator/package.json:
```diff
@@ -0,0 +1,25 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/arellaTV/lambda-megaphone-validator.git"
```
Were you going to change this to the scpr github org? You may be on the hook for maintaining this forever if it's pointed to your github.

arellaTV replied:
@nakedsushi, Great catch! Yes, I was going to change this to SCPR's github, thanks for the reminder. Should be fixed in the latest commit 🙌

nakedsushi commented on template.yaml:
```diff
@@ -0,0 +1,41 @@
+AWSTemplateFormatVersion: '2010-09-09'
```
What does this file do again?

arellaTV replied:
@nakedsushi This file allows you to test the function locally via `sam build` and `sam local start-api`

nakedsushi commented on README.md:
```diff
@@ -1,2 +1,8 @@
 # lambda-megaphone-validator
 An AWS lambda function that compares the file sizes of recently uploaded audio files with the original files
+
```
I'd add a note here about what environment variables are required here too, even though you have a .env.example

arellaTV replied:
@nakedsushi, Thank you for the suggestion! I went ahead and updated here as well.

nakedsushi reviewed (commented):
Have you thought about bubbling up errors to either have the lambda throw an error, or log the error with a magic string like `[ERROR]` or whatever, and then having CloudWatch monitor that and alert you (at your email address, or scprdev email) when there errors? That way, you won't need to remember to constantly check the Cloudwatch logs.

Basic info here:
http://marcelog.github.io/articles/aws_cloudwatch_monitor_lambda_alerts.html
https://stackoverflow.com/questions/42627341/send-notification-alert-when-aws-lambda-function-has-an-error

Lemme know if you want help setting it up.

arellaTV committed:
Exits early if a required env variable is missing, + updated docs

arellaTV commented:
Thanks for the review @nakedsushi! I'm going to continue to work on this PR, and document the results in this ticket: https://github.com/SCPR/digital_products/issues/959

arellaTV committed:
Adds tagged logs and fixes bug with UTC date

arellaTV commented:
Thanks again for the initialk review @nakedsushi! I've updated with you feedback, and added tagged logs.

The new logs are JSON strings, like `{ "level": "ERROR", "message": "404" }`, and a metric filter is set up with this expression `{ $.level = "ERROR" }`. I then set up an alarm that looks at this metric filter and sends an email to a notification group (which so far only includes my email, jarella@scpr.org). The alarm goes off if there is at least one error logged in 5 minutes.

nakedsushi commented:
Nice job!

nakedsushi commented on megaphone-validator/lib/reupload-audio.js:
```diff
@@ -21,9 +21,9 @@ module.exports = (megaphoneEpisode, audioUrl) => {
-            resolve(body);
+        request(options, (error, response, body) => {
+            if (error) reject({ level: "ERROR", message: error });
+            resolve({ level: "REUPLOADED", message: `${programSlug}'s audio file was reuploaded`, body });
```
Typically, the levels are like `INFO` `WARNING` or `ERROR`.  Maybe this one could be `WARNING` since it failed gracefully but might still be more "errory" than `INFO`?

arellaTV replied:
@nakedsushi 
Ah! Yes that makes complete sense. I'll go ahead and adjust this to `WARNING`, and then that way if we ever need warning for something else we can use that bucket for that too.

arellaTV committed:
Switches log level from REUPLOADED to WARNING

arellaTV commented:
Squashed, merged, and shipped to production
